llmfoundry/utils/config_utils.py

```diff
@@ -86,6 +86,25 @@ def process_init_device(model_cfg: DictConfig, fsdp_config: Optional[Dict]):
             # Set defaults for mixed initialization
             fsdp_config.setdefault('use_orig_params', False)
             fsdp_config.setdefault('load_monolith_rank0_only', True)
+
+    # no mixed precision needed for weights when they're already 16 bits
+    master_dtype = model_cfg.get('master_weights_dtype')
+    small_dtypes = ('bf16', 'f16', 'float16', 'bfloat16', 'amp_fp16',
+                    'amp_bf16')
+    if fsdp_config and master_dtype in small_dtypes:
+        reduce_dtype = None
+        buffer_dtype = None
+        mixed_precision = fsdp_config.get('mixed_precision')
+        if isinstance(mixed_precision, Mapping):
```

Comment:
what happens when people just use `mixed_precision: PURE` or `DEFAULT`?

Reply:
I had `mixed_precision: PURE` in my runs. All the mixed precision dtypes just default to `None` if the config isn't a mapping.

Reply:
iirc
if your runs used `mixed_precision: PURE`, then mixed_precision was not a mapping, then in your runs `reduce_dtype` and `buffer_dtype` defaulted to `None`, which means in FSDP they default to fp32, therefore you were not using PURE

Reply:
Do they default to fp32 specifically, or whatever dtype they are in the model? I have a separate `model.to(dtype=torch.bfloat16)` [in `train.py`](https://github.com/mosaicml/llm-foundry/blob/71e3f9c426e7490e1032a1e5a801bde07c295aeb/scripts/train/train.py#L222).

I *think* FSDP is using bf16 weights and grads when requested based on the ~6GB memory savings across 8 GPUs on a 7B model [in wandb](https://wandb.ai/mosaic-ml/lion8b-v2/reports/LION8b-MPT-Results--Vmlldzo1MTMzMTE5). (7e9 params * 6 bytes saved per param) / 8 devices = 5.25GB. And the marginal gain of ~3.7GB going from {4B grad, 4B weight, 1B opt state} -> {2B grad, 3B weight, 1B opt state} is consistent with saving 7e9/8 * 3bytes = 2.625GB.

It's actually a little weird to me that we're saving *more* memory than we'd expect, but maybe this is the pytorch allocator or FSDP not binpacking perfectly.

Reply:
I guess if its None, then param_dtype is used: https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/api.py#L125